include/alp/algorithms/householder_qr.hpp

```diff
@@ -0,0 +1,244 @@
+			const Divide &divide = Divide()
+		) {
+			RC rc = SUCCESS;
+
+			const Scalar< D > zero( ring.template getZero< D >() );
+			const Scalar< D > one( ring.template getOne< D >() );
+			const size_t n = nrows( H );
+			const size_t m = ncols( H );
+
+#ifdef DEBUG
+			std::cout << " n, m= " << n << ", " << m << "\n";
+#endif
+
+			// Q = identity( n )
+			rc = alp::set( Q, zero );
+			auto Qdiag = alp::get_view< alp::view::diagonal >( Q );
+			rc = rc ? rc : alp::set( Qdiag, one );
+			if( rc != SUCCESS ) {
+				std::cerr << " set( Q, I ) failed\n";
+				return rc;
+			}
+
+			// Out of place specification of the computation
+			Matrix< D, GeneralType, Dense > RR( n, m );
+
+			rc = set( RR, H );
+			if( rc != SUCCESS ) {
+				std::cerr << " set( RR, H ) failed\n";
+				return rc;
+			}
+#ifdef DEBUG
+			print_matrix( " << RR >> ", RR );
+#endif
+
+			// a temporary for storing the mxm result
+			Matrix< D, OrthogonalType, Dense > Qtmp( n, n );
+
+			for( size_t k = 0; k < std::min( n-1, m ); ++k ) {
+#ifdef DEBUG
+				std::string matname(" << RR(");
+				matname = matname + std::to_string(k);
+				matname = matname + std::string( ") >> ");
+				print_matrix( matname , RR );
+#endif
+
+ 				//const size_t m = n - k - 1;
+
+				// ===== Begin Computing v =====
+				// v = H[ k + 1 : , k ]
+				// alpha = norm( v ) * v[ 0 ] / norm( v[ 0 ] )
+				// v = v - alpha * e1
+				// v = v / norm ( v )
+
+				auto v_view = get_view( RR, utils::range( k, n ), k );
+				Vector< D, GeneralType, Dense > v( n - k );
+				rc = set( v, v_view );
+				if( rc != SUCCESS ) {
+					std::cerr << " set( v, view ) failed\n";
+					return rc;
+				}
+
+				Scalar< D > alpha( zero );
+				rc = norm2( alpha, v, ring );
+				if( rc != SUCCESS ) {
+					std::cerr << " norm2( alpha, v, ring ) failed\n";
+					return rc;
+				}
+
+				rc = eWiseLambda(
+					[ &alpha, &ring, &divide, &minus ]( const size_t i, D &val ) {
+						if ( i == 0 ) {
+							Scalar< D > norm_v0( std::abs( val ) );
+							Scalar< D > val_scalar( val );
+							foldl( alpha, val_scalar, ring.getMultiplicativeOperator() );
+							foldl( alpha, norm_v0, divide );
+							foldl( val_scalar, alpha, minus );
+							val = *val_scalar;
+						}
+					},
+					v
+				);
+				if( rc != SUCCESS ) {
+					std::cerr << " eWiseLambda( lambda, v ) failed\n";
+					return rc;
+				}
+
+				Scalar< D > norm_v( zero );
+				rc = norm2( norm_v, v, ring );
+				if( rc != SUCCESS ) {
+					std::cerr << " norm2( norm_v, v, ring ) failed\n";
+					return rc;
+				}
+
+				rc = foldl(v, norm_v, divide );
+#ifdef DEBUG
+				print_vector( " v = ", v );
+#endif
+				// ===== End Computing v =====
+
+				// ===== Calculate reflector Qk =====
+				// Q_k = identity( n )
+				typedef typename std::conditional<
+					grb::utils::is_complex< D >::value,
+					structures::Hermitian,
+					structures::Symmetric
+				>::type SymmOrHerm;
+				Matrix<	D, SymmOrHerm, Dense > Qk( n );
+				rc = alp::set( Qk, zero );
```

Comment:
Here `alp::` is used, but in most other places it is not. Consider either removing it here or, maybe better, adding it to other calls to alp primitives.